### PR TITLE
Fix failing Optimizer spec tests with ID generation

### LINKED ISSUE #4798: Failing tests in `ampproject/optimizer` library
## Bug Description

The tests in the `ampproject/optimizer` are currently failing due to the following recently merged PR in the `ampproject/amp-toolbox` package:

https://github.com/ampproject/amp-toolbox/pull/763

Link to failing build: https://travis-ci.org/github/ampproject/amp-wp/builds/692440360

The failures are due to the ID generation not using the same prefix and indexing scheme.

## Expected Behaviour

All tests against the spec files in `ampproject/amp-toolbox` should pass.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by schlessera**
## Summary

Fixes the spec test failures introduced by the merge of https://github.com/ampproject/amp-toolbox/pull/763 in the `ampproject/amp-toolbox` package.

- Changes the default prefix for `Document::getElementId()` from `i-amp-id` to `i-amp`
- Changes the indexing behavior for `Document::getElementId()` to start with `0` instead of `1`
- Changes the indexing behavior for `Document::getElementId()` to always add the index instead of skipping the first instance
- (fixed upstream) ~Skips the following tests until https://github.com/ampproject/amp-toolbox/issues/809 is resolved:~
	~- `'ServerSideRendering - converts_sizes_attribute_to_css'`~
	~- `'ServerSideRendering - converts_heights_attribute_to_css'`~
	~- `'ServerSideRendering - converts_media_attribute_to_css'`~
- Adds `CssRule` and `CssRules` abstractions so that CSS can be compacted (same properties merged into one rule with multiple selectors)
- Adapts the `ServerSideRendering` so that it makes use of `CssRule(s)` objects to adapt blocking attributes
- Fixes buggy media query negation logic
- Skips the following test until https://github.com/ampproject/amp-toolbox/issues/819 is resolved:
	- 'ServerSideRendering - converts_sizes_attribute_to_css'

Fixes #4798

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

**Comment by westonruter** [comment]
There's one remaining test failure:

```
1) AmpProject\Optimizer\SpecTest::testTransformerSpecFiles with data set "ServerSideRendering - converts_media_attribute_to_css" ('ServerSideRendering - convert...to_css', 'AmpProject\Optimizer\Transfor...dering', '<html ⚡>\n  <head>\n    <scri...tml>\n', '<html ⚡ i-amphtml-layout i-am.../html>')
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
-    21 => '@media not all and (min-width: 650px){#i-amp-0,#my-image,#i-amp-4{display:none}}@media not screen and (min-width: 650px){#i-amp-1{display:none}}@media screen and (min-width: 650px){#i-amp-2{display:none}}@media (min-width: 650px){#i-amp-3,#i-amp-5,#i-amp-6{display:none}}'
+    21 => '@media not all and (min-width: 650px){#i-amp-0{display:none}}@media not screen and (min-width: 650px){#i-amp-1{display:none}}@media not not screen and (min-width: 650px){#i-amp-2{display:none}}@media not not (min-width: 650px){#i-amp-3{display:none}}@media not not (min-width: 650px){#i-amp-5{display:none}}@media not not (min-width: 650px){#i-amp-6{display:none}}@media not all and (min-width: 650px){#my-image{display:none}}@media not all and (min-width: 650px){#i-amp-4{display:none}}'

/app/public/content/plugins/amp/lib/optimizer/tests/src/MarkupComparison.php:45
/app/public/content/plugins/amp/lib/optimizer/tests/src/MarkupComparison.php:65
/app/public/content/plugins/amp/lib/optimizer/tests/SpecTest.php:120
```

**Review by westonruter — CHANGES_REQUESTED** [review]
One more failing test.

File: lib/optimizer/tests/src/MarkupComparison.php [hunk 1]
```diff
@@ -23,10 +23,26 @@ protected function assertEqualMarkup($expected, $actual)
...
+                    $attrs
+                );
+            }
+            return $token;
+        };
+
+        $expected_tokens = array_map( $normalize_attributes, array_filter( preg_split( '#(<[^>]+>|[^<>]+)#', $expected, -1, PREG_SPLIT_DELIM_CAPTURE ) ) );
+        $actual_tokens   = array_map( $normalize_attributes, array_filter( preg_split( '#(<[^>]+>|[^<>]+)#', $actual, -1, PREG_SPLIT_DELIM_CAPTURE ) ) );
+
+        $this->assertEquals( $expected_tokens, $actual_tokens );
```
Comment (schlessera) [inline lib/optimizer/tests/src/MarkupComparison.php]: This should actually go into `assertSimilarMarkup()`, not `assertEqualMarkup()`.

Comment (westonruter) [reply lib/optimizer/tests/src/MarkupComparison.php]: Ok, feel free to revert.

Should `assertSimilarMarkup()` be used in the failing tests then?

Or should the underlying order of the attributes be fixed?

Comment (schlessera) [reply lib/optimizer/tests/src/MarkupComparison.php]: I'm already working on fixing everything. The first should be used where we want to enforce exact markup, the second where we just want to ensure the functionality stays the same.
Here, I'll use the second, as the logic is different from both packages in when they generate the ID. So the only thing that moves around is where in the attributes list the ID is defined.

**Review by westonruter — APPROVED** [review]
Looks perfect. Great job.

**Pull request merged**